spec/Section 6 -- Execution.md

```diff
@@ -42,44 +73,106 @@ of variable's declared type. If a query error is encountered during
+    *serially*.
+  * Let {errors} be any *field errors* produced while executing the
+    selection set.
+  * Return an unordered map containing {data} and {errors}.
+
+
+## Executing Selection Sets
 
+To execute a selection set, the object value being evaluated and the object type
+need to be known, as well as whether it must be executed serially, or may be
+executed in parallel.
 
-## Evaluating selection sets
+First, the selection set is turned into a grouped field set; then, each
+represented field in the grouped field set produces an entry into a
+response map.
 
-To evaluate a selection set, the executor needs to know the object on which it
-is evaluating the set and whether it is being evaluated serially.
+ExecuteSelectionSet(selectionSet, objectType, objectValue, variableValues):
 
-If the selection set is being evaluated on the `null` object, then the result
-of evaluating the selection set is `null`.
+  * Initialize {visitedFragments} to be the empty set.
+  * Let {groupedFieldSet} be the result of
+    {CollectFields(objectType, selectionSet, visitedFragments, variableValues)}.
+  * Initialize {resultMap} to an empty ordered map.
+  * For each {groupedFields} in {groupedFieldSet}:
+    * Let {entryTuple} be {GetFieldEntry(objectType, objectValue, groupedFields, variableValues)}.
+    * If {entryTuple} is not {null}:
+      * Let {responseKey} and {responseValue} be the values of {entryTuple}.
+      * Set {responseValue} as the value for {responseKey} in {resultMap}.
+  * Return {resultMap}.
 
-Otherwise, the selection set is turned into a grouped field set; each entry in
-the grouped field set is a list of fields that share a responseKey.
+Note: {responseMap} is ordered by which fields appear first in the query. This
+is explained in greater detail in the Response section below.
 
-The selection set is converted to a grouped field set by calling
-`CollectFields`, initializing `visitedFragments` to an empty list.
+Note: Normally, each call to {GetFieldEntry()} in the algorithm above is
+performed in parallel. However there are conditions in which each call must be
+done in serial, such as for mutations. This is explain in more detail in the
+sections below.
 
-CollectFields(objectType, selectionSet, visitedFragments):
+Before execution, the selection set is converted to a grouped field set by
+calling {CollectFields()}. Each entry in the grouped field set is a list of
+fields that share a response key.
+
+This ensures all fields with the same response key (alias or field name)
+included via referenced fragments are executed at the same time.
+
+CollectFields(objectType, selectionSet, visitedFragments, variableValues):
 
   * Initialize {groupedFields} to an empty ordered list of lists.
-  * For each {selection} in {selectionSet};
+  * For each {selection} in {selectionSet}:
     * If {selection} provides the directive `@skip`, let {skipDirective} be that directive.
```

Comment:
This actually isn't already a validation rule and is valid, though it does weird things. Changing this is probably not appropriate in an editing PR, but we should definitely fix this. 

Filed #223